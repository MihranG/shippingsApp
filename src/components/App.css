.app_wrapper {
    display: flex;
    flex-direction: column;
    align-items: center;
}

.app_header{
    display: flex;
    align-items: center;
}

.search_bar{
    width: 300px;
    margin: 0 10px
}