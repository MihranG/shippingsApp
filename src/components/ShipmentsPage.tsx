import * as React from 'react';

const ShipmentPage: React.FC<{}> = () =>{

    return(<div>
        
    </div>)
} 

export default ShipmentPage