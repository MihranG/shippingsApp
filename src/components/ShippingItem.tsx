import * as React from 'react'
import { Field, reduxForm } from 'redux-form'
import {TextField, Button, Paper, Typography, makeStyles, Theme, createStyles} from '@material-ui/core'
import { connect, compose } from 'react-redux'
import { IState } from '../interfaces'
import {RouteComponentProps} from 'react-router-dom'
import { editShipmentWithId } from '../store'


interface IComponentProps extends RouteComponentProps<{}> {

}

interface IStateProps {
    itemValues: any
    formValues: any
    editShippingItem: (id: number, shipmentDetails: any)=>void
}

interface IProps extends IComponentProps, IStateProps{
    
}


export const useStyles = makeStyles((theme: Theme) =>
  createStyles({
  
    paper: {
      marginBottom: theme.spacing(2),
      padding: 15
    },
    form: {
      width: 500
    },
    button: {
      margin: 15
    },
    type: {
      margin: '15px 0'
    },
    visuallyHidden: {
      border: 0,
      clip: 'rect(0 0 0 0)',
      height: 1,
      margin: -1,
      overflow: 'hidden',
      padding: 0,
      position: 'absolute',
      top: 20,
      width: 1,
    },
  }),
);

const validate = values => {
    const errors = {}
    const requiredFields = [
      'name',
      'destination',
      'mode',
      'origin',
      'type'
    ]
    requiredFields.forEach(field => {
      if (!values[field]) {
        errors[field] = 'Required'
      }
    })
    
    return errors
  }
// @ts-ignore
export const EditShipmentFormDisconnected: React.FC<IProps> = ({editShippingItem, match, formValues, itemValues , history, ...rest})=>{
    const handleFormSubmit = (e:React.FormEvent<HTMLFormElement>)=>{
        e.preventDefault();
        editShippingItem(match.params.id, formValues.values);
        history.replace('/')
    }

    const handleFormCancel = ()=>{
      history.replace('/')
    }

    const classes = useStyles();
    return (
        <form onSubmit={handleFormSubmit} className={classes.form}>
          <Paper elevation={3} className={classes.paper} >
            <Typography component='h3'> 
                  ID: {itemValues.id}
            </Typography>
            <div>
                <Field
                    name="name"
                    component={renderTextField}
                    label="First Name"
                />
            </div>
            <div>
                <Field
                    name="destination"
                    component={renderTextField}
                    label="Destination"
                />
            </div>
            <div>
                <Field
                    name="origin"
                    component={renderTextField}
                    label="Origin"
                />
            </div>
            <Typography component='h3' className={classes.type}> 
                Type: {itemValues.type}
            </Typography>
            <Button disabled={(formValues && (!formValues.anyTouched || !!formValues.syncErrors ))} type='submit' variant='contained' className={classes.button}> Change</Button>
            <Button type='reset' variant='text' onClick={handleFormCancel}> Cancel</Button>
          </Paper>
        </form>
    )
}


const renderTextField = ({
    label,
    input,
    meta: { touched, invalid, error },
    ...custom
  }) => (
    <TextField
      label={label}
      placeholder={label}
      error={touched && invalid}
      helperText={touched && error}
      {...input}
      {...custom}
    />
  )


const mapStateToProps :(state: IState, ownProps : RouteComponentProps)=>{itemValues: object, formValues: object} = (state, ownProps)=>{
  const itemValues = state.shipments.data[ownProps.match.params.id];
  return({
      itemValues,
      formValues: state.form.editShipping,
      initialValues: itemValues
    })
  }

const mapDispatchToProps =  (dispatch) =>({
    editShippingItem: (id, shipmentDetails)=>dispatch(editShipmentWithId({id, shipmentDetails}))
})



export default connect(mapStateToProps, mapDispatchToProps)(
  reduxForm({
    form: 'editShipping',
    validate
  })(EditShipmentFormDisconnected)
)
