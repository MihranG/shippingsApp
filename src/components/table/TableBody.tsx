import * as React from 'react';
import { connect } from 'react-redux';
import { createStyles, makeStyles, Theme } from '@material-ui/core/styles';
import {Table, TableBody,TableCell, TablePagination, TableRow, Paper, LinearProgress, TableContainer} from '@material-ui/core';

import TableHeader from './TableHeader';
import Header from '../HeaderComponent';
import {Order, IShipment, IState} from '../../interfaces';
import { start } from 'repl';


function stableSort<T>(array: T[], cmp: (a: T, b: T) => number, page: number, rowsPerPage: number) {

  const startIndex = page * rowsPerPage;
  const endIndex = startIndex + rowsPerPage;
  const stabilizedTuple= array.slice(startIndex, endIndex ).map((el, index) => [el, index] as [T, number]);
  stabilizedTuple.sort((a, b) => {
    const order = cmp(a[0], b[0]);
    if (order !== 0) return order;
    return a[1] - b[1];
  });
  const sortedArray = [
    ...array.slice(0,startIndex),
    ...stabilizedTuple.map(el => el[0]),
    ...array.slice(endIndex)
  ]

  return sortedArray
}



function getSorting<K extends keyof any>(
  order: Order,
  orderBy: K,
): (a: { [key in K]: number | string }, b: { [key in K]: number | string }) => number {
  const isAscending: boolean = order === 'asc' ;
  const isAscendingMultiplier: number = isAscending ? -1 : 1
  return function (a,b){
    if (b[orderBy] < a[orderBy]) {
      return -1 * isAscendingMultiplier;
    }
    if (b[orderBy] > a[orderBy]) {
      return 1 * isAscendingMultiplier;
    }
    return 0;
  }
  
}

export const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      width: '100%',
    },
    paper: {
      width: '100%',
      marginBottom: theme.spacing(2),
    },
    table: {
      minWidth: 750,
    },
    visuallyHidden: {
      border: 0,
      clip: 'rect(0 0 0 0)',
      height: 1,
      margin: -1,
      overflow: 'hidden',
      padding: 0,
      position: 'absolute',
      top: 20,
      width: 1,
    },
  }),
);
function TableBodyDisconnected({shippings, history, isLoading}) {
    const classes = useStyles();
    const [orderType, setOrderType] = React.useState<Order>('asc');
    const [orderBy, setOrderBy] = React.useState<keyof IShipment>('name');
    const [page, setPage] = React.useState(0);
    const [rowsPerPage, setRowsPerPage] = React.useState(20);

    const handleRequestSort = (event: React.MouseEvent<unknown>, property: keyof IShipment) => {
      const isAsc = orderBy === property && orderType === 'asc';
      setOrderType(isAsc ? 'desc' : 'asc');
      setOrderBy(property);
    };
  
  
    const handleChangePage = (event: unknown, newPage: number) => {
      setPage(newPage);
    };
  
    const handleChangeRowsPerPage = (event: React.ChangeEvent<HTMLInputElement>) => {
      setRowsPerPage(parseInt(event.target.value, 10));
      setPage(0);
    };
    
    const emptyRows = rowsPerPage - Math.min(rowsPerPage, shippings.length - page * rowsPerPage);
    const arrayOfAppearingRows = stableSort(shippings, getSorting(orderType, orderBy), page, rowsPerPage)
                                    .slice(page * rowsPerPage, page * rowsPerPage + rowsPerPage)

    return (
      <>
       <Header  history={history}/>
        {isLoading ? <LinearProgress variant="query" color="secondary" />
        :(<Paper className={classes.paper}>
          <TableContainer>
            <Table
              className={classes.table}
              aria-labelledby="tableTitle"
              size={'small'}
              aria-label="enhanced table"
            >
              <TableHeader
                classes={classes}
                orderType={orderType}
                orderBy={orderBy}
                onRequestSort={handleRequestSort}
                rowCount={shippings.length}
              />
              <TableBody>
              
                {arrayOfAppearingRows.map((row, index) => {
                    const labelId = `enhanced-table-checkbox-${index}`;
                    return (
                      <TableRow
                        hover
                        onClick={(a: React.MouseEvent<HTMLTableRowElement, MouseEvent>):void=>{
                            history.push(`/edit/${row.id}`)}
                        }
                        role="checkbox"
                        tabIndex={-1}
                        key={row.id}
                      >
                        <TableCell component="th" id={labelId} scope="row" align='left'>
                          {row.id}
                        </TableCell>
                        <TableCell align="left">{row.name}</TableCell>
                        <TableCell align="right">{row.destination}</TableCell>
                        <TableCell align="right">{row.origin}</TableCell>
                        <TableCell align="right">{row.type}</TableCell>
                      </TableRow>
                    );
                  })}
                {emptyRows > 0 && (
                  <TableRow style={{ height: (33) * emptyRows }}>
                    <TableCell colSpan={6} />
                  </TableRow>
                )}
              </TableBody>
            </Table>
          </TableContainer>
          <TablePagination
            rowsPerPageOptions={[20, 40]}
            component="div"
            count={shippings.length}
            rowsPerPage={rowsPerPage}
            page={page}
            onChangePage={handleChangePage}
            onChangeRowsPerPage={handleChangeRowsPerPage}
          />
        </Paper>)}
        </>
    );
  }
  
  const mapStateToProps = (state: IState)=>({
    shippings : Object.values(state.shipments.data),
    isLoading: state.shipments.isLoading,
  })
  export default connect(mapStateToProps)(TableBodyDisconnected)