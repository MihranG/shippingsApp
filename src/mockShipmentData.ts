import {IShipment} from './interfaces'

export const data: IShipment[]  =
    [{
        "id": "S1000",
        "name": "TEST1",
        "cargo": [
          {
            "type": "Fabric",
            "description": "1000 Blue T-shirts",
            "volume": "2"
          },
          {
            "type": "Fabric",
            "description": "2000 Green T-shirts",
            "volume": "3"
          }
        ],
        "mode": "sea",
        "type": "FCL",
        "destination": "Saarbrücker Str. 38, 10405 Berlin",
        "origin": "Shanghai Port",
        "services": [
          {
            "type": "customs"
          }
        ],
        "total": "1000",
        "status": "ACTIVE",
        "userId": "U1000"
      },
      {
        "id": "S1001",
        "name": "TEST2",
        "cargo": [
          {
            "type": "Furniture",
            "description": "300 Tables",
            "volume": "20"
          },
          {
            "type": "Furniture",
            "description": "1500 Chairs",
            "volume": "15"
          }
        ],
        "mode": "sea",
        "type": "FCL",
        "destination": "Saarbrücker Str. 38, 10405 Berlin",
        "origin": "Ningbo port",
        "services": [
          {
            "type": "customs"
          },
          {
            "type": "insurance",
            "value": "100"
          }
        ],
        "total": "3000",
        "status": "ACTIVE",
        "userId": "U1002"
      }
  ]

export default data