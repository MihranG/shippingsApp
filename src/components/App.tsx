import * as React from "react";
import {Dispatch} from 'redux';
import { hot } from "react-hot-loader";
import {Provider, connect} from 'react-redux';
import {store} from '../store'
import { fetchAllData } from "../thunk";
import TableBody  from "./table/TableBody";
import { BrowserRouter , Switch, Route } from 'react-router-dom'
import ShippingItem from "./ShippingItem";
import NotFound from './NotFound';

import './App.css'
import { IShipment } from "../interfaces";

//const reactLogo = require("./../assets/img/react_logo.svg");
//import "./../assets/scss/App.scss";


interface IProps {
    getShipments: ()=>Promise<void>
    shipments: IShipment[]
}

export class AppDisconnected extends React.Component<IProps, undefined> {
    componentDidMount(){
        this.props.getShipments()
    }

    handleShipmentResultClick(){
        this.props
    }

    public render() {
        const {shipments} = this.props
        return (
            <div className='app_wrapper'>
                <Provider store={store}>
                    <BrowserRouter>
                        <Switch>
                            <Route exact path="/" component={TableBody}/>
                            <Route path="/edit/:id" component={ShippingItem}/>
                            <Route path="*" component= {NotFound}/>                      
                        </Switch>
                    </BrowserRouter >
                </Provider>
            </div>
        );
    }
} 

declare let module: object;

const mapStateToProps = (state)=>({
    page: state.pages,
    shipments: Object.values(state.shipments.data)
})

const mapDispatchToProps = (dispatch: Dispatch):{
    getShipments: any
} =>{
    return {
        //@ts-ignore
        getShipments: dispatch(fetchAllData())
    }
}
const ConnectedApp = connect(mapStateToProps, mapDispatchToProps)(AppDisconnected);

export const Root = () => <Provider store={store}><ConnectedApp/></Provider>

export default hot(module)(Root);
