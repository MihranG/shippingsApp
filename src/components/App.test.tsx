import { shallow, mount } from 'enzyme';
import {Root, AppDisconnected} from './App';
import {data as mockShipmentsData} from '../mockShipmentData';
import * as React from 'react';
import TableRow from '@material-ui/core/TableRow';
import ShipmentsPage from './HeaderComponent'
import {BrowserRouter} from 'react-router-dom'
import ShipmentPage from './HeaderComponent';
import ShippingItem from './ShippingItem';
import TableBody from './table/TableBody';
import { TableHead } from '@material-ui/core';


describe('App component shallow', ()=>{
    let wrapper;
    
    const getShipments = jest.fn((x)=>x)
    beforeAll(()=>{
        // @ts-ignore
        wrapper = shallow(<AppDisconnected  getShipments= {getShipments}/>,);
    }),
    test('should match snapshot',()=>{
        expect(wrapper).toMatchSnapshot()
    })
    test('should call getShipments() methid in first load', ()=>{
        expect(getShipments.mock.calls.length).toEqual(1)
    })

    afterAll(()=>{
        wrapper.detach()
    })
    
})

describe('All app',()=>{
    let wrapper;
    beforeAll(()=>{
        // const mockResponse = mockData;
        const mockJsonPromise = Promise.resolve(mockShipmentsData);
        const mockFetchPromise = Promise.resolve({ // 3
            json: () => mockJsonPromise,
          });
        // @ts-ignore
        global.fetch = jest.fn().mockImplementation(() => mockFetchPromise);
    })
   
    beforeAll(()=>{
        wrapper = mount(<Root />)
    })

    test('should match snapshot', ()=>{
        expect(wrapper).toMatchSnapshot()
    })

    test('should depict table with appropriate initial routing', ()=>{
        const router = wrapper.find(BrowserRouter);
        const table = wrapper.find(TableBody);
        const rowQty = table.find(TableHead)
        const itemPage = wrapper.find(ShippingItem);
        
        expect(router.length).toBe(1);
        expect(table.length).toBe(1);
        expect(itemPage.length).toBe(0);
        expect(rowQty.length).toBe(0);
    })

    afterAll(()=>{
        // @ts-ignore
        global.fetch.mockClear();
        // @ts-ignore
        delete global.fetch;
        wrapper.detach()
    })
})