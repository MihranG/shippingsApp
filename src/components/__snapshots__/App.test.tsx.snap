// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`All app should match snapshot 1`] = `ReactWrapper {}`;

exports[`App component shallow should match snapshot 1`] = `ShallowWrapper {}`;
